sdk/client/src/otel/index.ts

```diff
@@ -0,0 +1,319 @@
+import { XMLHttpRequestInstrumentation } from '@opentelemetry/instrumentation-xml-http-request'
+import { Resource } from '@opentelemetry/resources'
+import { OTLPTraceExporter } from '@opentelemetry/exporter-trace-otlp-http'
+import {
+	SEMRESATTRS_DEPLOYMENT_ENVIRONMENT,
+	SEMRESATTRS_SERVICE_NAME,
+} from '@opentelemetry/semantic-conventions'
+import { BatchSpanProcessorBase } from '@opentelemetry/sdk-trace-base/build/src/export/BatchSpanProcessorBase'
+import * as api from '@opentelemetry/api'
+import {
+	BrowserXHR,
+	getBodyThatShouldBeRecorded,
+} from '../listeners/network-listener/utils/xhr-listener'
+import type { NetworkRecordingOptions } from '../types/client'
+import { sanitizeHeaders } from '../listeners/network-listener/utils/network-sanitizer'
+import { shouldNetworkRequestBeTraced } from '../listeners/network-listener/utils/utils'
+import { UserInteractionInstrumentation } from './user-interaction'
+
+export type OtelConfig = {
+	projectId: string | number
+	sessionSecureId: string
+	endpoint?: string
+	environment?: string
+	networkRecordingOptions?: NetworkRecordingOptions
+	serviceName?: string
+	tracingOrigins?: boolean | (string | RegExp)[]
+}
+
+let provider: WebTracerProvider
+
+export const initializeOtel = (config: OtelConfig) => {
+	if (provider !== undefined) {
+		console.warn('OTEL already initialized. Skipping...')
+		return
+	}
+
+	// TODO: Is there a better way to determine if we're in dev mode? Seems
+	// import.meta.env.DEV is set to dev when building with dev:frontend.
+	const isDev = window.location.hostname === 'localhost'
+	const fallbackEndpoint = isDev
+		? 'https://localhost:4318'
+		: 'https://otel.highlight.io:4318'
+
+	const endpoint = config.endpoint ?? fallbackEndpoint
+	const environment = config.environment ?? 'production'
+
+	provider = new WebTracerProvider({
+		resource: new Resource({
+			[SEMRESATTRS_SERVICE_NAME]:
+				config.serviceName ?? 'highlight-browser',
+			[SEMRESATTRS_DEPLOYMENT_ENVIRONMENT]: environment,
+			'highlight.project_id': config.projectId,
+			'highlight.session_id': config.sessionSecureId,
+		}),
+	})
+
+	provider.register({
+		propagator: new CompositePropagator({
+			propagators: [
+				new W3CBaggagePropagator(),
+				new W3CTraceContextPropagator(),
+			],
+		}),
+	})
+
+	// Export spans to console for debugging
+	if (isDev) {
+		provider.addSpanProcessor(
+			new SimpleSpanProcessor(new ConsoleSpanExporter()),
+		)
+	}
+
+	const exporter = new OTLPTraceExporter({
+		url: endpoint + '/v1/traces',
+		concurrencyLimit: 3,
+		// Using any because we were getting an error importing CompressionAlgorithm
+		// from @opentelemetry/otlp-exporter-base.
+		compression: 'gzip' as any,
+	})
+
+	const spanProcessor = new CustomSpanProcessor(exporter, {
+		tracingOrigins: config.tracingOrigins,
+	})
+	provider.addSpanProcessor(spanProcessor)
+
+	// Can customize by passing a config object as 2nd param.
+	provider.addSpanProcessor(new BatchSpanProcessor(exporter))
```

Comment:
any tweaks needed to the config, like we do on the backend sdks to batch larger?

Reply:
I was going to stick with the defaults for now, but wanted to make it known we could customize things like batch size if we decide we need to. That sound alright to you?

Reply:
yup!